addons/udes_stock/models/stock_picking.py

```diff
@@ -2598,6 +2598,71 @@ def search_for_pickings(self, picking_type_id, picking_priorities, limit=1, doma
+        Location = self.env["stock.location"]
+        MoveLine = self.env["stock.move"]
+        Quant = self.env["stock.quant"]
+
+        # Create empty record sets for SO lines
+        unfulfillable_lines = MoveLine.browse()
+
+        _logger.info("Checking pickings to cancel due to stock shortage")
+        # Get unreserved stock for each product in locations
+        locations = Location.get_available_stock_locations()
+        stock = defaultdict(int)
+        skip_states = ("assigned", "done", "cancel")
+
+        for r, batch in self.batched(size=batch_size):
+            _logger.info("Checking pickings %d-%d", r[0], r[-1])
+            # Cache the needed fields and only the needed fields
+            # This code has to process tens of thousands of sale order lines
+            # and hundreds of thousands of stock moves.
+            # Caching too much is expensive here because of the sheer number of
+            # records processed: Odoo's field loading becomes a bottleneck and
+            # memory usage skyrockets.
+            # Caching too little is also expensive since:
+            #  * cache misses cause unused fields to be loaded in due to
+            #    prefetching, which leads to overcaching, described above
+            #  * cache misses for stock move fields result in hundreds of small
+            #    loads per batch, which is inefficient due to overheads
+            # For non-relational fields, read() is used instead of mapped()
+            # because it allows for loading of specific fields, whereas
+            # mapped() will load in as many fields as it can due to prefetching.
+            # with_context(prefetch_fields=False) could be used with mapped()
+            # but is limited to a single column at a time.
+            batch.mapped("move_lines")
+            batch.mapped("move_lines").read(
+                ["product_id", "product_uom_qty", "state"], load="_classic_write"
+            )
+
+            for picking in batch:
+                # Loop lines and deduct from stock dict, add move lines to
+                # can or can't fulfill record sets.
+                # If this code is modified, the caching above needs to be
+                # kept up to date to ensure good performance
+                for line in picking.move_lines.filtered(lambda m: m.state not in skip_states):
+
+                    product = line.product_id
+
+                    if product not in stock.keys():
+                        stock[product] = Quant.get_available_quantity(product, locations)
+                    qty_ordered = line.product_uom_qty
+                    if stock[product] <= 0 or (
+                        stock[product] < qty_ordered
+                        and not picking.picking_type_id.u_handle_partials
+                    ):
+                        unfulfillable_lines |= line
+                    stock[product] -= qty_ordered
+
+            # Empty cached stuff
+            batch.invalidate_cache()
```

Comment:
Not sure about this cache invalidation, since the follow up code could benefit from what we have cached when there is enough stock.